crates/biome_service/src/dome.rs

```diff
@@ -0,0 +1,82 @@
+use biome_fs::BiomePath;
+use rustc_hash::FxHashSet;
+use std::collections::hash_set::Iter;
+use std::iter::Peekable;
+
+#[derive(Debug, Default)]
+/// A type that holds the evaluated paths, and provides an iterator to extract
+/// specific paths like configuration files, manifests and more.
+pub struct Dome {
+    paths: FxHashSet<BiomePath>,
+}
+
+impl Dome {
+    pub fn with_path(mut self, path: impl Into<BiomePath>) -> Self {
+        self.paths.insert(path.into());
+        self
+    }
+
+    pub fn new(paths: FxHashSet<BiomePath>) -> Self {
+        Self { paths }
+    }
+
+    pub fn iter(&self) -> DomeIterator {
+        DomeIterator {
+            iter: self.paths.iter().peekable(),
+        }
+    }
+
+    pub fn to_paths(self) -> FxHashSet<BiomePath> {
+        self.paths
+    }
+}
+
+pub struct DomeIterator<'a> {
+    iter: Peekable<Iter<'a, BiomePath>>,
+}
+
+impl<'a> DomeIterator<'a> {
+    pub fn next_config(&mut self) -> Option<&'a BiomePath> {
+        return if let Some(path) = self.iter.peek() {
+            if path.is_config() {
+                self.iter.next()
+            } else {
+                None
+            }
+        } else {
+            None
+        };
+    }
+
+    pub fn next_ignore(&mut self) -> Option<&'a BiomePath> {
+        return if let Some(path) = self.iter.peek() {
+            if path.is_ignore() {
+                self.iter.next()
+            } else {
+                None
+            }
+        } else {
+            None
+        };
+    }
+
+    pub fn next_manifest(&mut self) -> Option<&'a BiomePath> {
+        return if let Some(path) = self.iter.peek() {
+            if path.is_manifest() {
+                self.iter.next()
+            } else {
+                None
+            }
+        } else {
+            None
+        };
+    }
```

Comment:
I am unsure to get the idea behind these methods.

They are used in `biome_cli::execute::traverse.rs`:

```rs
    fs.traversal(Box::new(|scope: &dyn TraversalScope| {
        while let Some(path) = iter.next_config() {
            scope.handle(ctx, path.to_path_buf());
        }

        while let Some(path) = iter.next_manifest() {
            scope.handle(ctx, path.to_path_buf());
        }

        for path in iter {
            scope.handle(ctx, path.to_path_buf());
        }
    }));
```

Does this mean that they are yield in-order? i.e. first the configs, then the manifest, and finally the others?

If it is the case, then I don't understand how it is possible because `FxHashSet` has no order guarantee? [BTreeSet](https://doc.rust-lang.org/std/collections/struct.BTreeSet.html#) could be use instead if the order is required.

Moreover, if they are meant to be yielded in order, then we could use a single `next()` and use loop that iterate until a change in the file kind is detected. This could avoid these methods, and we could remove `Peekable` that has some cost.

Reply:
These are just proposals, they are "used" in a sense that they don't break existing test.

The idea is to have this command throw errors instead, e.g., `try_next_manifest`, which will throw an error if we don't consume the configuration files first.

These methods will enforce an order on how files should be processed. The workflow I have in mind:
- detect configuration files, and create a "project" for each one of them (very much like we do in the LSP for workspace projects)
- detect manifest files. If they belong to a project, update it, otherwise we store a new project that doesn't have a configuration file 
- detect ignore files
- handle and inspect files
- handle files

Something like that

> [BTreeSet](https://doc.rust-lang.org/std/collections/struct.BTreeSet.html#) could be use instead if the order is required.

Yeah I wasn't sure of it, I will apply the change